dra: scheduler queueing hints

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Scheduling with less overhead and without delays when ResourceClaims are involved.

#### Which issue(s) this PR fixes:

Related-to: https://github.com/kubernetes/kubernetes/issues/118893#issuecomment-1608942940

#### Special notes for your reviewer:

This is a combination of two related enhancements:

- By implementing a PreEnqueue check, the initial pod scheduling
  attempt for a pod with a claim template gets avoided when the claim
  does not exist yet.
- By implementing cluster event checks, only those pods get
  scheduled for which something changed, and they get scheduled
   immediately without delay.

#### Does this PR introduce a user-facing change?
```release-note
faster scheduling when ResourceClaims are involved
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3063
```

pohly committed:
scheduler/dra: reduce pod scheduling latency

This is a combination of two related enhancements:
- By implementing a PreEnqueue check, the initial pod scheduling
  attempt for a pod with a claim template gets avoided when the claim
  does not exist yet.
- By implementing cluster event checks, only those pods get
  scheduled for which something changed, and they get scheduled
  immediately without delay.

pohly committed:
scheduler dynamicresources: minor logging improvements

This makes some complex values a bit more readable.

pohly commented:
/assign @sanposhiho

sanposhiho commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,256 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+		{Event: framework.ClusterEvent{Resource: framework.PodSchedulingContext, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterPodSchedulingContextChange},
 		// A resource might depend on node labels for topology filtering.
 		// A new or updated node may make pods schedulable.
 		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.UpdateNodeLabel}},
```
Cannot we add QueueingHint on Node add event/update event? (Or do you want to do it in another PR?)

pohly replied:
It should be doable, but definitely in a separate PR. Getting the checks wrong is a potential source of bugs, so I don't want to do too much at once.

sanposhiho reviewed (commented):
I haven't got enough time to review it, will do the full review later

carlory commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,256 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+}
+
+// isSchedulableAfterClaimChange is invoked whenever a claim changed. It checks whether
+// that change made a previously unschedulable pod schedulable. It errs on the side of
```
```suggestion
// that change made a previous unschedulable pod schedulable. It errs on the side of
```

pohly replied:
`Previously` refers to `unschedulable`, not `pod`. I believe `previously` is correct here - but I am not a native speaker.

sanposhiho commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,256 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+		return framework.QueueSkip
+	}
+
+	// Deleted? That can happen because we ourselves delete the PodSchedulingContext while
+	// working on the pod. This can be ignored.
+	if oldObj != nil && newObj == nil {
+		pl.logger.V(4).Info("PodSchedulingContext for pod got deleted", "pod", klog.KObj(pod))
+		return framework.QueueSkip
+	}
```
This deleted event handling can be done around L372 (before converting type).

pohly replied:
The log message then shouldn't mention the pod. It's simpler, so I moved the check.

sanposhiho commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,256 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+	// resources on the selected node become available again. It's
+	// much more likely, in particular when trying to fill up the
+	// cluster, that the choice simply didn't work out. The risk
+	// here is that in a situation where the cluster really is
+	// full, backoff won't be used because the scheduler keeps
+	// trying different nodes. This should not happen when it has
+	// full knowledge about resource availability (=
+	// PodSchedulingContext.*.UnsuitableNodes is complete) but may happen
+	// when it doesn't (= PodSchedulingContext.*.UnsuitableNodes had to be
+	// truncated).
+	if podScheduling.Spec.SelectedNode != "" {
```
I'd like to understand this part correctly.

Let's say Node1 - 4 are in the cluster and Node2-4 doesn't have enough capacity. So, the scheduler set `podScheduling.Spec.SelectedNode=Node1`. But, later, the device driver said `claimStatus.UnsuitableNodes = Node1`. And we reach here and the Pod will get QueueImmediately.

In this case, in the next scheduling cycle, Node2-4 are still rejected by the resource fit plugin or whatever, and Node1 is rejected by the DRA plugin. Meaning, this Pod won't be schedulable until Node2-4 get free space.
But, the Pod could keep getting rescheduled by QueueImmediately.

You mentioned such risk in the above comment, right? 
How can we avoid this situation? Could you elaborate `This should not happen when it has full knowledge about resource availability`?

pohly replied:
> But, the Pod could keep getting rescheduled by QueueImmediately.

There will only be one scheduling attempt, the one triggered by this cluster event. After that, there won't be any further updates of the `PodSchedulingContext` unless a driver reports a change in resource availability, so there will be no further "immediate" queueing.

> You mentioned such risk in the above comment, right?

That risk is slightly different: `UnsuitableNodes` has a size limit. If a driver cannot report all unsuitable nodes, then the scheduler might alternate between different "selected nodes" that are all unsuitable.

This is an issue that we need to keep an eye on, but I think it can be ignored for now in this PR (alpha stage, not limited to scheduler events).

sanposhiho replied:
> There will only be one scheduling attempt, the one triggered by this cluster event. 

Ah, true.

> UnsuitableNodes has a size limit. If a driver cannot report all unsuitable nodes, then the scheduler might alternate between different "selected nodes" that are all unsuitable.

OK, that's what you said "truncated". Understood.

sanposhiho replied:
Can we have a comment to describe the risk is acceptable for now?

pohly replied:
Added. I also added a few more comments in other places.

sanposhiho commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,256 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+		}
+	}
+
+	if oldPodScheduling != nil &&
```
If we reach here, `oldPodScheduling != nil` is always true due to L442.

pohly replied:
True, changed.

sanposhiho replied:
Maybe you have forgotten to fix this?

pohly replied:
No, `oldPodScheduling` is nil for a "create" event.

sanposhiho replied:
Nah, OK never mind. 🙏

sanposhiho commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,256 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+		return framework.QueueSkip
+	}
+
+	if loggerV := pl.logger.V(6); loggerV.Enabled() {
+		loggerV.Info("PodSchedulingContext for pod with no relevant changes, maybe schedule", "pod", klog.KObj(pod), "podSchedulingDiff", cmp.Diff(oldPodScheduling, podScheduling))
+	} else {
+		pl.logger.V(5).Info("PodSchedulingContext for pod with no relevant changes, maybe schedule", "pod", klog.KObj(pod))
+	}
+	return framework.QueueAfterBackoff
```
In which case we could reach here? Cannot we return QueueSkip if we know it's non relevant change?

pohly replied:
We don't know for sure that the change can be ignored. When in doubt, "backoff" (and thus try later) IMHO is the better choice and matches what the scheduler would do without scheduling hints.

I'll change the wording to "with unknown changes" instead of "no relevant changes", that's more accurate for this situation. The diff is included at `V(6)` precisely because a developer might want to know more.

sanposhiho replied:
That makes sense

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,256 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+		return framework.QueueSkip
+	}
+
+	modifiedClaim, ok := newObj.(*resourcev1alpha2.ResourceClaim)
```
Looking... `As` doesn't handle `cache.DeletedFinalStateUnknown`. That is okay here (can't occur for `newObj`, but for `oldObj` in `isSchedulableAfterPodSchedulingContextChange` it is relevant.

Shall I add support for it in this PR?

pohly replied:
I've added it, will be in the next push.

sanposhiho replied:
Thanks!

pohly committed:
scheduler util: handle cache.DeletedFinalStateUnknown in As

Informer callbacks must be prepared to get cache.DeletedFinalStateUnknown as
the deleted object. They can use that as hint that some information may have
been missed, but typically they just retrieve the stored object inside it.

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,256 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+	return nil
+}
+
+// isSchedulableAfterClaimChange is invoked whenever a claim changed. It checks whether
```
... or deleted. Instead of listing all possible reasons, I changed this to "is invoked for all claim events reported by an informer".

This also explains why handling `cache.DeletedFinalStateUnknown` is required.

pohly commented:
@sanposhiho : I pushed [an update](https://github.com/kubernetes/kubernetes/compare/09496db88b902bc54cc07bc27a8f8654f2ea12e2..28d3ae4922c96798effc828c5415e8c680b0e693), please take another look.

sanposhiho commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,235 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+		loggerV.Info("PodSchedulingContext for pod with unknown changes, maybe schedule", "pod", klog.KObj(pod), "podSchedulingDiff", cmp.Diff(oldPodScheduling, podScheduling))
+	} else {
+		pl.logger.V(5).Info("PodSchedulingContext for pod with unknown changes, maybe schedule", "pod", klog.KObj(pod))
+	}
```
Can we have some comment to describe that if we reach here, something unknown happens and that's why we return QueueAfterBackoff

pohly replied:
Done.

sanposhiho commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:
```diff
@@ -838,3 +873,213 @@ func createReactor(tracker cgotesting.ObjectTracker) func(action cgotesting.Acti
+		},
+	}
+
+	for name, tc := range testcases {
```
nit: do you want to make this test parallel as well? (because other UTs in this test do that.

pohly replied:
Done.

sanposhiho reviewed (commented):
Only nits

pohly commented:
@sanposhiho: [updated](https://github.com/kubernetes/kubernetes/compare/28d3ae4922c96798effc828c5415e8c680b0e693..e76fcab7b37f8fe7359f678a4089180bacc25177) once more.

sanposhiho reviewed (commented):
Everything is clear to me. Thanks!
/lgtm

sanposhiho commented:
@kubernetes/sig-scheduling-leads @kerthcet 
Could anyone else give another review for `/approve` on it?

sanposhiho commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -208,28 +212,42 @@ func statusForClaim(schedulingCtx *resourcev1alpha2.PodSchedulingContext, podCla
 
 // New initializes a new plugin and returns it.
 func New(plArgs runtime.Object, fh framework.Handle, fts feature.Features) (framework.Plugin, error) {
+	// TODO: the runtime should set up logging for each plugin, including
+	// adding a name for each one (same as in kube-controller-manager).
+	return NewWithLogger(klog.TODO(), plArgs, fh, fts)
```
Do we have an issue for this TODO by the way?

pohly replied:
Not specifically for this one, but there are PRs pending for the conversion to contextual logging. Once those are complete, we can search for klog.TODO and fix it if not done yet.

kerthcet replied:
Seems better to pass-in a logger here, rather than a context like this https://github.com/kubernetes/kubernetes/pull/116884/files#diff-b5d92dd0897babb368df8c1aad9e95c957eca378bd46481d6d8501011f722715R77-R81

pohly replied:
Whatever gets done here must get coordinated when converting kube-scheduler to contextual logging. Let's not open that rathole right now.

carlory commented on pkg/scheduler/util/utils.go:
```diff
@@ -164,6 +165,8 @@ func IsScalarResourceName(name v1.ResourceName) bool {
 // Both objects must be of the same type. If not, an error is returned.
 // nil objects are allowed and will be converted to nil.
+// For oldObj, cache.DeletedFinalStateUnknown is handled and the
+// pobject stored in it will be converted instead.
```
```suggestion
// pobject stored in it will be converted instead.
// pod object stored in it will be converted instead.
```

kerthcet commented:
I will take a look tomorrow, I still have jobs at hand right now.

kerthcet commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,248 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+
+	if !usesClaim {
+		// This was not the claim the pod was waiting for.
+		pl.logger.V(6).Info("unrelated claim got modified", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim))
```
why not still v(4), I think it's similar with other logs.

pohly replied:
Because this is going to much more common than the other ones and not as interesting as those. If a developer wants to know everything, they should use -v6.

pohly replied:
Some other examples:
```
pkg/kubelet/pleg/evented.go:			if klog.V(6).Enabled() {
pkg/kubelet/pleg/evented.go-				klog.ErrorS(err, "Evented PLEG: error generating pod status from the received event", "podUID", podID, "podStatus", status)
--
pkg/kubelet/pleg/evented.go:			if klogV := klog.V(6); klogV.Enabled() {
pkg/kubelet/pleg/evented.go-				klogV.InfoS("Evented PLEG: Generated pod status from the received event", "podUID", podID, "podStatus", status)
--
pkg/kubelet/pleg/generic.go:		if klog.V(6).Enabled() {
pkg/kubelet/pleg/generic.go-			klog.ErrorS(err, "PLEG: Write status", "pod", klog.KRef(pod.Namespace, pod.Name), "podStatus", status)
--
pkg/kubelet/pleg/generic.go:		if klogV := klog.V(6); klogV.Enabled() {
pkg/kubelet/pleg/generic.go-			klogV.InfoS("PLEG: Write status", "pod", klog.KRef(pod.Namespace, pod.Name), "podStatus", status)
--
staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:	if klogV := klog.V(6); klogV.Enabled() {
staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go-		chosenQueue := qs.queues[bestQueueIdx]
```

kerthcet commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,248 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+	}
+
+	pl.logger.V(4).Info("status of claim for pod got updated", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim))
+	return framework.QueueImmediately
```
I think we should be strict about `QueueImmediately`, I didn't see any special reason here, do we have?

pohly replied:
It is needed to make the scheduler respond without backoff delay when a claim got allocated.

sanposhiho replied:
QueueImmediately in this case should be OK.
Previous discussion: https://github.com/kubernetes/kubernetes/pull/117561#discussion_r1186573820

kerthcet commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,248 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+	oldPodScheduling, newPodScheduling, err := schedutil.As[*resourcev1alpha2.PodSchedulingContext](oldObj, newObj)
+	if err != nil {
+		// Shouldn't happen.
+		pl.logger.Error(nil, "isSchedulableAfterPodSchedulingChange")
```
Let's keep the same with `isSchedulableAfterClaimChange` as `isSchedulableAfterPodSchedulingContextAddOrUpdate`?

kerthcet commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,248 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+		return framework.QueueSkip
+	}
+
+	if oldPodScheduling == nil /* create */ ||
```
One question about `AllocationMode`, is it immutable?

pohly replied:
Yes.

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,248 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+func (pl *dynamicResources) PreEnqueue(ctx context.Context, pod *v1.Pod) (status *framework.Status) {
+	logger := klog.FromContext(ctx)
+	defer func() {
+		logger.V(5).Info("PreEnqueue", "pod", klog.KObj(pod), "status", status)
```
Done.

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -251,22 +269,248 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+	podScheduling := newPodScheduling // Never nil because deletes are handled above.
+
+	if podScheduling.Name != pod.Name || podScheduling.Namespace != pod.Namespace {
+		pl.logger.V(7).Info("PodSchedulingContext for unrelated pod got modified", "pod", klog.KObj(pod), "podScheduling", klog.KObj(podScheduling))
```
Nothing >5 is standardized. Each additional increment just adds more information. Therefore `V(6)` seems fine to me. `V(10)` tends to be extremely noisy, which would make debugging this code at `-v10` not very practical.

pohly replied:
See also https://github.com/kubernetes/kubernetes/pull/119078#discussion_r1258440448 for other examples.

pohly commented:
/retest

sanposhiho commented:
/retest

sanposhiho reviewed (approved):
/lgtm
/approve

sanposhiho commented:
/retest
